gpMgmt/bin/gprebalance_modules/shrink.py

```
@@ -0,0 +1,509 @@
+        },
+        {
+            'trigger': 'move_to_STATE_END',
+            'source': ['STATE_SHRINK_DONE', 'STATE_CHECK_PREVIOUS_RUN'],
+            'dest': 'STATE_END'
+        },
+        {
+            'trigger': 'move_to_STATE_ERROR',
+            'source': '*',
+            'dest': 'STATE_ERROR'
+        }
+    ]
+
+    def __init__(self, logger, dburl, options, gpEnv, gpArray, gpArrayDumpFilename):
+        self.logger = logger
+        self.dburl = dburl
+        self.options = options
+        self.gpEnv = gpEnv
+        self.conn = dbconn.connect(
+            self.dburl, encoding='UTF8', allowSystemTableMods=True)
+        self.shutdown_requested = False
+        self.workers_for_tables_rebalance = None
+        self.workers_for_segment_stop = None
+        self.gparray = gpArray
+        self.gparray_dump_file = gpArrayDumpFilename
+
+        self.machine = Machine(model = self,
+                               queued=True,
+                               states = self.states + self.states_main_shrink_flow,
+                               transitions = self.transitions,
+                               initial = 'STATE_START',
+                               before_state_change = 'on_every_state')
+
+    def run(self):
+        self.trigger('start')
+
+    def ggrebalance_schema_exists(self) -> bool:
+        with closing(dbconn.connect(self.dburl, encoding='UTF8')) as conn:
+            row = dbconn.queryRow(conn, f"SELECT COUNT(1) FROM pg_namespace WHERE nspname = '{self.rebalance_schema_name}'")
+            return int(row[0]) == 1
+        return False
+
+    def get_state_from_previous_run(self) -> str:
+        with closing(dbconn.connect(self.dburl, encoding='UTF8')) as conn:
+            cursor = dbconn.query(conn, f'SELECT status FROM {self.rebalance_schema_name}.{self.rebalance_status} order by updated DESC limit 1')
+            if cursor.rowcount > 0:
+                return str(cursor.fetchone()[0])
+        return 'not defined'
```

Is this valid return value? Maybe throw an error or assert that gprebalance schema exists and has a valid state name or otherwise it's been manually tampered with?

---

Having this value here allows to get more or less meaningfull error message like
"Previous run stopped after state 'not defined', trying to continue from the next state..." (with going to error state after it), meaning that either the rebalance schema was manually changed, or there is some error in gpshrink logic.
Sophisticated error handling is not in scope of this skeleton commit.

---

Maybe add a check that is triggered in `on_every_state()` or use enums instead of strings for state, so it's more type/lint-safe?

---

What kind of check are you suggesting? Please elaborate a bit more.

---

```python
assert self.machine.get_state() in self.states + self.states_main_shrink_flow
```

---

Added assert. Regarding switching to enums - currently I do not like this idea, as it will make code for recovery from an interrupted state more complicated.